internal/checker/nodebuilder.go

```diff
@@ -0,0 +1,2988 @@
+	truncating     bool
+	addedLength    int
+	trackedSymbols []*TrackedSymbolArgs
+}
+
+type CompositeTypeCacheIdentity struct {
+	typeId        TypeId
+	flags         nodebuilder.Flags
+	internalFlags nodebuilder.InternalFlags
+}
+
+type NodeBuilderLinks struct {
+	serializedTypes                  map[CompositeTypeCacheIdentity]*SerializedTypeEntry // Collection of types serialized at this location
+	fakeScopeForSignatureDeclaration *string                                             // If present, this is a fake scope injected into an enclosing declaration chain.
+}
+
+type NodeBuilderSymbolLinks struct {
+	specifierCache module.ModeAwareCache[string]
+}
+type NodeBuilderContext struct {
+	tracker                         nodebuilder.SymbolTracker
+	approximateLength               int
+	encounteredError                bool
+	truncating                      bool
+	reportedDiagnostic              bool
+	flags                           nodebuilder.Flags
+	internalFlags                   nodebuilder.InternalFlags
+	depth                           int
+	enclosingDeclaration            *ast.Node
+	enclosingFile                   *ast.SourceFile
+	inferTypeParameters             []*Type
+	visitedTypes                    map[TypeId]bool
+	symbolDepth                     map[CompositeSymbolIdentity]int
+	trackedSymbols                  []*TrackedSymbolArgs
+	mapper                          *TypeMapper
+	reverseMappedStack              []*ast.Symbol
+	enclosingSymbolTypes            map[ast.SymbolId]*Type
+	suppressReportInferenceFallback bool
+	remappedSymbolReferences        map[ast.SymbolId]*ast.Symbol
+
+	// per signature scope state
+	hasCreatedTypeParameterSymbolList     bool
+	hasCreatedTypeParametersNamesLookups  bool
+	typeParameterNames                    map[TypeId]*ast.Identifier
+	typeParameterNamesByText              map[string]struct{}
+	typeParameterNamesByTextNextNameCount map[string]int
+	typeParameterSymbolList               map[ast.SymbolId]struct{}
+}
+
+type NodeBuilder struct {
+	// host members
+	f  *ast.NodeFactory
+	ch *Checker
+	e  *printer.EmitContext
+
+	// cache
+	links       core.LinkStore[*ast.Node, NodeBuilderLinks]
+	symbolLinks core.LinkStore[*ast.Symbol, NodeBuilderSymbolLinks]
+
+	// closures
+	typeToTypeNodeClosure               func(t *Type) *ast.TypeNode
+	typeReferenceToTypeNodeClosure      func(t *Type) *ast.TypeNode
+	conditionalTypeToTypeNodeClosure    func(t *Type) *ast.TypeNode
+	createTypeNodeFromObjectTypeClosure func(t *Type) *ast.TypeNode
+	isStringNamedClosure                func(d *ast.Declaration) bool
+	isSingleQuotedStringNamedClosure    func(d *ast.Declaration) bool
+
+	// state
+	ctx *NodeBuilderContext
+}
+
+const (
+	defaultMaximumTruncationLength      = 160
+	noTruncationMaximumTruncationLength = 1_000_000
+)
+
+// Node builder utility functions
+
+// You probably don't mean to use this - use `NewNodeBuilderAPI` instead
+func NewNodeBuilder(ch *Checker, e *printer.EmitContext) NodeBuilder {
```

Comment:
If this (and the `NodeBuilder` struct altogether) are not to be used directly, does it need to be exported? If this were `nodeBuilder`, then `NodeBuilderInterface` could become `NodeBuilder`. All the small structs above in this file also seem not to escape the `checker` package either.

Reply:
My goal was to move the implementation into the `nodebuilder` package by abstracting the `Checker` APIs it uses, but turns out the builder uses over 120(!) checker APIs - which isn't itself an issue beyond the effort involved, but they also require moving `Type`/`Signature`/`IndexInfo` (and all the derived subtypes of `Type`) up the stack, too (and, in turn, making their fields public or into methods), which I know I'll get... _opinions_ on, so I'm saving that for a followup. I can lowercase it for now... but it doesn't matter either way, since it's in the `internal` package. Honestly, I don't know why we wouldn't default to package exported in this environment.

Reply:
Yeah, I think the reasons to keep stuff unexported are just for us, not for future importers. It keeps completions in other packages cleaner and faster, and my understanding is it’s good for whole project builds (you can edit signatures of unexported stuff without invalidating the build of downstream packages). But mainly having to name interfaces `FooInterface` makes me sad 😄